### PR TITLE
Auto select Credit card type based on number

### PR DESCRIPTION
**Pull request by arunnethues**
Asana Card
https://app.asana.com/0/676656131310897/1200794831818869/f

### CODE REVIEW

File: checkout/components/user-payment.js [hunk 1]
```diff
@@ -60,5 +60,19 @@ app.component('user-payment', {
...
+    changeCardType (event) {
+      let visa = new RegExp('^4[0-9]{12}(?:[0-9]{3})?$');
+      let mastercard = new RegExp('^5[1-5][0-9]{14}$');
+      let mastercard2 = new RegExp('^2[2-7][0-9]{14}$');
+      let cc = event.target.value;
+      let card_type_drop_down= document.getElementById('card_type_drop_down');
+      if (visa.test(cc)) {
+        card_type_drop_down.value ="visa";
+      } else if (mastercard.test(cc) || mastercard2.test(cc)) {
+        card_type_drop_down.value ="master";
```
Comment (gaurav-bains) [inline checkout/components/user-payment.js:74]:
```suggestion
      if (visa.test(cc)) {
        this.paymentDetails.cardType ="visa";
      } else if (mastercard.test(cc) || mastercard2.test(cc)) {
        this.paymentDetails.cardType ="master";
```
@arunnethues VueJS works on data reactivity, in the version you created, only value of dropdown was changed but the data was not being updated
test using this code on local, and check response in console, first console log will have `this.paymentDetails.cardType` as "" but second console log will show the correct value, I have suggested the code above that updates the data but input selection happens as VueJS being reactive changes the dropdown selection. Does it make sense? If not, I can explain in detail as well
```changeCardType (event) {
      let visa = new RegExp('^4[0-9]{12}(?:[0-9]{3})?$');
      let mastercard = new RegExp('^5[1-5][0-9]{14}$');
      let mastercard2 = new RegExp('^2[2-7][0-9]{14}$');
      let cc = event.target.value;
      let card_type_drop_down= document.getElementById('card_type_drop_down');
      if (visa.test(cc)) {
        card_type_drop_down.value ="visa";
      } else if (mastercard.test(cc) || mastercard2.test(cc)) {
        card_type_drop_down.value ="master";
      }
      console.log(this.paymentDetails);
      this.paymentDetails.cardType = card_type_drop_down.value;
      console.log(this.paymentDetails);
    }
```

**Review by gaurav-bains — CHANGES_REQUESTED** [review]

File: checkout/components/user-payment.js [hunk 2]
```diff
@@ -5,7 +5,7 @@ app.component('user-payment', {
   template: `
   <div>
     <label>Card Type: *</label>
-    <select v-model.trim="paymentDetails.cardType">
+    <select v-model.trim="paymentDetails.cardType" id="card_type_drop_down">
```
Comment (gaurav-bains) [inline checkout/components/user-payment.js:8]:
```suggestion
    <select v-model.trim="paymentDetails.cardType">
```

File: checkout/components/user-payment.js [hunk 3]
```diff
@@ -14,7 +14,7 @@ app.component('user-payment', {
   </div>
   <div>
     <label>Card Number: *</label>
-    <input type="text" maxlength="16" v-model.trim="paymentDetails.cardNumber">
+    <input type="text" maxlength="16" v-model.trim="paymentDetails.cardNumber" @change="changeCardType($event)">
```
Comment (gaurav-bains) [inline checkout/components/user-payment.js:17]:
```suggestion
    <input type="text" maxlength="16" v-model.trim="paymentDetails.cardNumber" @change="changeCardType()">
```

File: checkout/components/user-payment.js [hunk 4]
```diff
@@ -60,5 +60,19 @@ app.component('user-payment', {
     expYears() {
       return Array.from({length:10},(v,k)=>k+ new Date().getFullYear())
     }
+  },
+  methods: {
+    changeCardType (event) {
+      let visa = new RegExp('^4[0-9]{12}(?:[0-9]{3})?$');
+      let mastercard = new RegExp('^5[1-5][0-9]{14}$');
+      let mastercard2 = new RegExp('^2[2-7][0-9]{14}$');
+      let cc = event.target.value;
```
Comment (gaurav-bains) [inline checkout/components/user-payment.js:69]:
```suggestion
      let cc = paymentDetails.cardNumber;
```

File: checkout/components/user-payment.js [hunk 5]
```diff
@@ -60,5 +60,19 @@ app.component('user-payment', {
     expYears() {
       return Array.from({length:10},(v,k)=>k+ new Date().getFullYear())
     }
+  },
+  methods: {
+    changeCardType (event) {
```
Comment (gaurav-bains) [inline checkout/components/user-payment.js:65]:
```suggestion
    changeCardType () {
```

File: checkout/components/user-payment.js [hunk 6]
```diff
@@ -60,5 +60,19 @@ app.component('user-payment', {
...
+      let visa = new RegExp('^4[0-9]{12}(?:[0-9]{3})?$');
+      let mastercard = new RegExp('^5[1-5][0-9]{14}$');
```
Comment (pjdarch) [inline checkout/components/user-payment.js:67]: You can simplify/consolidate the Mastercard into a single regular expression. This reduces the need of running two checks

`'^(5[1-5]|2[2-7])[0-9]{14}$`

**Comment by arunnethues** [comment]
@gaurav-bains , @pjdarch 
thanks for suggestion, code is updated, please review

File: checkout/components/user-payment.js [hunk 7]
```diff
@@ -60,5 +60,17 @@ app.component('user-payment', {
...
+  },
+  methods: {
+    changeCardType () {
+      let visa = new RegExp('^4[0-9]{12}(?:[0-9]{3})?$');
+      let mastercard = new RegExp('^(5[1-5]|2[2-7])[0-9]{14}$');
+      let cc = this.paymentDetails.cardNumber;
+      if (visa.test(cc)) {
+        this.paymentDetails.cardType ="visa";
+      } else if (mastercard.test(cc)) {
+        this.paymentDetails.cardType ="master";
```
Comment (pjdarch) [inline checkout/components/user-payment.js:72]: I know there isn't any linting enforced on the project, but the style has been to single quotes.

```suggestion
        this.paymentDetails.cardType = 'visa';
      } else if (mastercard.test(cc)) {
        this.paymentDetails.cardType = 'master';
```

**Comment by pjdarch** [comment]
This is just me being partly OCD, as it isn't part of the original PR.. Can you also update L#45? `MasterCard` should be `Mastercard`

`cardTypes: {master: 'MasterCard', visa: 'VISA'}`

**Comment by arunnethues** [comment]
@pjdarch above changes has been done

**Review by pjdarch — APPROVED** [review]

**Review by gaurav-bains — APPROVED** [review]
eyeballed, looks good

**Pull request merged**